### PR TITLE
5746: Bugfix add timestamp to module assets

### PR DESCRIPTION
**Pull request by alfredbez**
- [x] add DocBlocks
- [x] add new Unit-Tests
- [x] rebase/squash

This fixed [0005746: Timestamp is not added for css and js files included from module](https://bugs.oxid-esales.com/view.php?id=5746)

### CODE REVIEW

**Comment by nymo** [comment]
I like this feature. I would propose to add the timestamp feature aswell for the css files in a module folder.

File: source/Core/ViewHelper/JavaScriptRegistrator.php [hunk 1]
```diff
@@ -60,45 +66,47 @@ public function addSnippet($script, $isDynamic = false)
...
-        $config = oxRegistry::getConfig();
         $suffix = $isDynamic ? '_dynamic' : '';
         $filesParameterName = static::FILES_PARAMETER_NAME . $suffix;
-        $includes = (array) $config->getGlobalParameter($filesParameterName);
+        $includes = (array) $this->config->getGlobalParameter($filesParameterName);
 
-        if (!preg_match('#^https?://#', $file)) {
-            $file = $this->formLocalFileUrl($file);
+        $isSameBaseUrlRegex = '#^' . preg_replace('^https?:/', 'https?:', $this->config->getCurrentShopUrl(false)) . '#';
+        $isSameBaseUrl = preg_match($isInternUrlRegex, $file);
```
Comment (alfredbez) [inline source/Core/ViewHelper/JavaScriptRegistrator.php]: use `$isSameBaseUrlRegex` instead of `$isInternUrlRegex`

**Comment by alfredbez** [comment]
> Current threshold is set to 659, but the amount of errors is 662.

I have no idea why the build failed:
```
vagrant@oxideshop:/var/www/oxideshop$ git rev-parse HEAD
164fe20542682ca34cd7db15f84219452cf8d821
vagrant@oxideshop:/var/www/oxideshop$ vendor/bin/phpcsoxid --report=summary source/Core/ViewHelper/*.php
vagrant@oxideshop:/var/www/oxideshop$
```

I'll try to close & reopen to trigger a new build.

**Comment by alfredbez** [comment]
@nymo and @keywan-ghadami-oxid done
If there's something missing, please let me know. Do we need additional Unit-Tests?

**Comment by keywan-ghadami-oxid** [comment]
Hi Alfred, thank you very much, as far i know there is a internal guideline that new features must always have unit tests. So i guess my colleges would be very happy if you can provide unit tests. As they have limited time slots for taking care about pull requests, it may help to get more pull requests accepted.

**Comment by nymo** [comment]
Hi Alfred,
thanks for integrating also the stylesheets into this feature! Looking forward for the merge :)

**Comment by alfredbez** [comment]
I added some unit tests

File: source/Core/ViewHelper/BaseRegistrator.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+        if (empty($url) && $this->config->getConfigParam('iDebug') != 0) {
+            $error = "{oxscript} resource not found: " . oxStr::getStr()->htmlspecialchars($url);
+            trigger_error($error, E_USER_WARNING);
+        }
+
+        return $url ? "$url?$parameters" : '';
+    }
+
+    public function getFileCreationTime($file) {
```
Comment (alfredbez) [inline source/Core/ViewHelper/BaseRegistrator.php]: missing DocBlock

**Comment by alfredbez** [comment]
I'd like to move `isSameBaseUrl` and `getPathByUrl` to `UtilsUrl`. What do you think?

**Comment by alfredbez** [comment]
This PR is ready for review. If there is anything else you need, please, let me know.

**Comment by alfredbez** [comment]
FYI: I resolved the merge conflicts

**Comment by alfredbez** [comment]
I'd be glad to help if I can do so.

**Comment by alfredbez** [comment]
Just stumbled upon this in a project again, would be awesome if we can have this in the core.

**Comment by Sieg** [comment]
Hey @alfredbez! Sorry for having this open for so long :( I would like to ask you to rebase this PR on our master, what do you think?

**Comment by alfredbez** [comment]
Done

**Comment by Sieg** [comment]
Hey, have checked a little - Couldnt isCurrentShopHost method be used in place of that newly introduced isSameBaseUrl?

**Comment by alfredbez** [comment]
I have to dig in again, since I have no idea what I have written 4 years ago :wink: 

I'm currently working on too many things, not sure when I'll be able to pick this up, but I expect not in the next weeks/months.

If you thing that it makes sense and it works, feel free to change my code, `Allow edits by maintainers` is checked.

**Comment by Sieg** [comment]
Hey @alfredbez, i have spent some more time on this one, cleaned up a bit - tried to understand all the changes done, and as i see, it requires much more time to make something out of it.

What we have here currently:

* some abstraction extracted for those registrators
  - not sure if its enough to do like this, still looks like the code 
* more complicated calculation of local path to css files for faster calculation of file timestamp based on replacing some path parts to find out the local file and get its time locally byt not via http?

But still:

* ViewConfig::getModuleUrl still not pointing to the new code in any way, so the bug linked is not solved by thins one as i understand.
* All main asset files got timestamps calculated without these changes, so no point to run it through more complicated (configuration dependant/unstable?) path calculations.

Maybe i am missing something, but looks like this PR should be closed, otherwise - much more tested and finalized to actually solve the bug tracker issue.

Please use pr493 branch for further improvements.

**Comment by Sieg** [comment]
Hey guys. Thanks for the solution. Merged with some finalization from my side at 4bcb59efa050e7e2fa696c80398e1740620ecf02. Let me know if i have missed something, or if we broke something! :) Looks fine, tested a bit, CI green.

**Pull request closed without merging**